src/Microsoft.TestPlatform.Client/Discovery/DiscoveryRequest.cs

```diff
@@ -355,7 +374,16 @@ public void HandleLogMessage(TestMessageLevel level, string message)
         /// <param name="rawMessage">Raw message.</param>
         public void HandleRawMessage(string rawMessage)
         {
-            this.OnRawMessageReceived?.Invoke(this, rawMessage);
+            var message = this.dataSerializer.DeserializeMessage(rawMessage);
+
+            if (!string.Equals(MessageType.DiscoveryComplete, message.MessageType))
+            {
+                this.OnRawMessageReceived?.Invoke(this, rawMessage);
+            }
+            else
+            {
+                this.discoveryCompleteTask = Task.Run(() => SendRawMessageAsync(rawMessage));
```

Comment:
We can avoid the additional Task if we just store the `rawMessage` content and send it just before the `runCompletionEvent` is set.

Reply:
I do want to DiscoveryRequest or TestRunRequest to cache messages so they can send them later.  Their responsibility is to send the messages, but not to store them.